### PR TITLE
Filter data count mismatch > Create the get_attribute_and_meta_counts method

### LINKED ISSUE #8563: Filter data count mismatch > Create the get_attribute_and_meta_counts method
To solve the filter data count mismatch problem as described in the epic https://github.com/woocommerce/woocommerce/issues/42466 , we need to create a new method to be responsible for manipulating the data received from the product meta and attribute queries (segments 4, 5 and 6, to be developed on woocommerce/woocommerce-blocks#8561 and woocommerce/woocommerce-blocks#8562) and return the final count, as represented by segments 1, 2 and 3 of the main query:

```sql
# Format final result (placing zero on null values)
SELECT coalesce(term_count, 0) as term_count, attributes.term_id as term_count_id
FROM (
         # Fetch all possible attribute values for a taxonomy.
         SELECT DISTINCT term_id
         FROM wp_wc_product_attributes_lookup
         WHERE taxonomy = 'pa_color') as attributes # 'pa_color' here corresponds to the filter by attribute color block.
         LEFT JOIN (

    # Count the products by attribute.
    SELECT COUNT(product_attribute_lookup.product_id) as term_count, product_attribute_lookup.term_id
    FROM wp_wc_product_attributes_lookup product_attribute_lookup
             INNER JOIN wp_posts posts
                        ON posts.ID = product_attribute_lookup.product_id

    # Ensure only published products are included.
    WHERE posts.post_type IN ('product', 'product_variation')
      AND posts.post_status = 'publish'

      # This is where the filter conditions are controlled.

      # The next segment corresponds to an "Any" (or) condition for attributes.
      # The filtered taxonomies in this example are blue (22) and green (23).
      # {start_condition_any_query}
      AND product_attribute_lookup.product_or_parent_id IN (
        # SEGMENT 4 - The output of the `get_filtered_product_terms` method
      # {end_condition_any_query}

      # The next segment corresponds to an "All" (and) condition for attributes.
      # The filtered taxonomies in this example are medium (26) and small (27)
      # {start_condition_all_query}
      AND product_attribute_lookup.product_or_parent_id IN (
        # SEGMENT 5 - The output of the `get_filtered_product_terms` method
      # {end_condition_all_query}

      # The next segment corresponds to a product meta filter.
      # The filtered meta in this example is the price with a max value of 18.
      # Other possible valid values are the stock and the rating.
      # {start_product_meta_query}
      AND product_attribute_lookup.product_id IN (
        # SEGMENT 6 - The output of the `get_filtered_product_metas` method
      # {end_product_meta_query}

    GROUP BY product_attribute_lookup.term_id
) summarize ON attributes.term_id = summarize.term_id
```

This macro query is the one responsible for ensuring the returned values are compared, counted, and correctly formatted. It is the one that, in this scenario, represents the updated version of our [get_attribute_counts](https://github.com/woocommerce/woocommerce-blocks/blob/cef076a63513e1595bbc2f19a5804187b8a41b51/src/StoreApi/Utilities/ProductQueryFilters.php/#L120).

The output of this function should be the same as the pre-existing `get_attribute_counts`: `array termId=>count pairs`

We should rely on pre-existing functionality from the core of Woo whenever possible and follow the best WordPress practices when structuring the new method and queries.

This method can be initially placed within the [ProductQueryFilters](https://github.com/woocommerce/woocommerce-blocks/blob/cef076a63513e1595bbc2f19a5804187b8a41b51/src/StoreApi/Utilities/ProductQueryFilters.php/#L9) class, with the possibility to be migrated upstream later on to the core of Woo (to be worked as part of a separate issue).

As soon as the implementation is complete, we should extensively test the filters on the FE with different combinations to confirm everything works as expected.

### LINKED ISSUE #8604: Filter data count mismatch > Ensure a single request is made to return the correct attributes count
While working on #8599 it was possible to verify that 3 API requests are made to return the attribute count data, but only the third one contains the term slug variable required to correctly calculate the results. As a consequence, in addition to fixing the queries, we also need to ensure the request is triggered once and only when needed.

### PR DESCRIPTION
**Pull request by nefeline**
In this PR, the new `get_attribute_and_meta_counts` method is introduced, which is responsible for fetching the counts for the Filter by Attribute block.

🎯  #8563 

These queries were tailor-made to handle all combinations of filters natively supported by WooCommerce blocks, including:

- Filter by Attribute
- Filter by stock status
- Filter by rating
- Filter by price

With that said, the following use-case scenarios were taken into consideration during the development and tests phase:

#### 1. A single Filter by Attribute block used alongside the Products (Beta) block with condition = any (`or`)

https://user-images.githubusercontent.com/15730971/229168789-2989c8ca-50cf-46e7-979b-f498612aa5f8.mov

#### 2. A single Filter by Attribute block used alongside the Products (Beta) block with condition = all (`and`)

https://user-images.githubusercontent.com/15730971/229170241-dda91be3-07e3-417c-8939-07274f5523a8.mov

#### 3. Two or more Filter by Attribute blocks used alongside the Products (Beta) block

https://user-images.githubusercontent.com/15730971/229168189-0844ca73-11fe-4ee4-ac16-32d31d1c2321.mov

#### 4. **Filter by Attribute block** used alongside the **Filter by stock status** and Products (Beta) blocks.

https://user-images.githubusercontent.com/15730971/229166865-61348b22-c471-4acc-b3f9-047bf715293e.mov

#### 5. **Filter by Attribute block** used alongside the **Filter by rating** and Products (Beta) blocks.

https://user-images.githubusercontent.com/15730971/229166103-1df378ac-8024-48f8-8b94-f0d793c8437e.mov

#### 6. **Filter by Attribute block** used alongside the **Filter by price** and Products (Beta) blocks.

https://user-images.githubusercontent.com/15730971/229165330-da29a7a1-6c78-4819-941d-4ba94ee06be7.mov

#### 7. **Filter by Attribute block** used alongside the **Filter by stock** status, Filter by price, Filter by rating, and Products (Beta) blocks.

https://user-images.githubusercontent.com/15730971/229163101-001d84bc-b9b0-4ad6-ab39-5600d2555a0f.mov

## Discussion points

- Should we update the counts when filtering attributes with the condition = any ( `or` )? Currently, the counts are updated to match the filtered product results. Alternatively, we could keep the counts unchanged. 

- Should we hide products with the count = zero when the condition is all ( `and` )? Currently, the filtered products with counts equal to zero are intentionally hidden. Alternatively, we could keep them on the list independently of the results.

Any preferences/thoughts?

## Note regarding PHP unit tests

I'm opening a separate PR for updating & fixing those.

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [x] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

#### User Facing Testing

- With a block theme such as twenty-twenty-three enabled, head over to Edit Site > Templates > Product Catalog > Edit
- Add the Products (Beta) block
- Add the Filter by attribute block
- Save and head over to your shop page
- Ensure the filter is properly rendered, and when filtering the products, the counts always match the displayed results
- Test all 7 combinations of filters as described and demonstrated in the description of this PR, ensure all of them work as expected.

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Fix: The product counts for the filter by attribute block now match the displayed products with different filter combinations, including filter by price, by rating, by stock, and by other attributes as well.

### CODE REVIEW

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (roykho) [inline src/StoreApi/Utilities/ProductQueryFilters.php:122]: Since this was a public function, perhaps we should deprecate the function instead of just replacing it? Otherwise people who are already using this function will get a breaking change.

Comment (roykho) [reply src/StoreApi/Utilities/ProductQueryFilters.php:122]: We previously discussed to add a caching layer in this function since we're touching the database multiple times in a query, it will speed things up.

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (roykho) [inline src/StoreApi/Utilities/ProductQueryFilters.php:142]: We're removing these main query filters is because we're no longer using the main query is that correct?

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (roykho) [inline src/StoreApi/Utilities/ProductQueryFilters.php]: Is the `-0.5` and `+0.5` needed? mysql BETWEEN should signify `=<` and `>=` from what I know.

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (roykho) [inline src/StoreApi/Utilities/ProductQueryFilters.php]: Perhaps just the `!empty()` will do here?

**Review by roykho — CHANGES_REQUESTED** [review]
This is the initial first round of review. I have left some inline comments.

In addition, when combining Filter by Attributes block and Filter by Ratings block, an error could occur.

https://user-images.githubusercontent.com/2132595/229878936-f232dda0-ace7-4403-babc-ce2b2e32b1e0.mp4

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (nefeline) [reply src/StoreApi/Utilities/ProductQueryFilters.php:142]: Yes, that's correct 👍

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (nefeline) [reply src/StoreApi/Utilities/ProductQueryFilters.php:122]: > We previously discussed to add a caching layer in this function since we're touching the database multiple times in a query, it will speed things up.

Good call, updated!

Comment (nefeline) [reply src/StoreApi/Utilities/ProductQueryFilters.php:122]: > Since this was a public function, perhaps we should deprecate the function instead of just replacing it? Otherwise people who are already using this function will get a breaking change.

That's a good call: instead of deprecating, I'm preserving the same original name for this method to prevent any problems with backward compatibility.

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (nefeline) [reply src/StoreApi/Utilities/ProductQueryFilters.php]: 👍 To define the value of the average rating, WooCommerce uses [PHP's `round()` function](https://www.php.net/manual/en/function.round.php) with `PHP_ROUND_HALF_UP`, so here I've updated the query to behave the same way, not including the upper bound value in this case.

**Review by nefeline — COMMENTED** [review]
Thank you so much for the review @roykho ! All feedback was addressed.

Regarding:

> In addition, when combining Filter by Attributes block and Filter by Ratings block, an error could occur.

I couldn't reproduce it locally, but I suspect that [this commit](https://github.com/woocommerce/woocommerce-blocks/pull/8599/commits/c98d1ff2e8962cab8c36a9cb50efa1af26ffae26) might have fixed the problem.

In case the issue persists, can you please let me know what you see in your PHP error log?

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (roykho) [reply src/StoreApi/Utilities/ProductQueryFilters.php:122]: I think it would still cause an issue because you have changed the function signature. I was thinking just use the same practice Woo core does when deprecating. Like this https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-emails.php#L308-L319

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (roykho) [reply src/StoreApi/Utilities/ProductQueryFilters.php]: I guess I am still not 100% sure the half up/down is needed? `BETWEEN` is inclusive already.

Comment (nefeline) [reply src/StoreApi/Utilities/ProductQueryFilters.php]: > I guess I am still not 100% sure the half up/down is needed? BETWEEN is inclusive already.

Yes, it's true that BETWEEN is inclusive, which is a problem; as an example: if the average rating is precisely 2.5, with `BETWEEN`, both the ratings 2 and 3 would be counted, but actually, we want just the 3 to be counted because that's the default behavior of the `round()` function. 

As a very basic example:

Screenshot 2023-04-04 at 18 01 54

**Comment by roykho** [comment]
Here is the PHP error regarding the `Rating Filter` example I showed in the video.

`[05-Apr-2023 12:45:58 UTC] WordPress database error You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ',2 - 0.5 AND average_rating < 5,2 + 0.5' at line 1 for query SELECT DISTINCT product_id FROM wp_wc_product_meta_lookup WHERE average_rating >= 5,2 - 0.5 AND average_rating < 5,2 + 0.5 made by require('wp-blog-header.php'), wp, WP->main, WP->parse_request, do_action_ref_array('parse_request'), WP_Hook->do_action, WP_Hook->apply_filters, rest_api_loaded, WP_REST_Server->serve_request, WP_REST_Server->dispatch, WP_REST_Server->respond_to_request, Automattic\WooCommerce\StoreApi\Routes\V1\AbstractRoute->get_response, Automattic\WooCommerce\StoreApi\Routes\V1\AbstractRoute->get_response_by_request_method, Automattic\WooCommerce\StoreApi\Routes\V1\ProductCollectionData->get_route_response, Automattic\WooCommerce\StoreApi\Utilities\ProductQueryFilters->get_attribute_counts, Automattic\WooCommerce\StoreApi\Utilities\ProductQueryFilters->get_product_by_metas`

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (roykho) [reply src/StoreApi/Utilities/ProductQueryFilters.php]: Interesting. When I tested, if I use `2.5`, I don't get `2` also.

**Comment by nefeline** [comment]
> Here is the PHP error regarding the Rating Filter example I showed in the video.

Thanks for sharing the error! Very helpful to narrow down the cause: the problem was solved for [the ratings](https://github.com/woocommerce/woocommerce-blocks/pull/8599/commits/97199212987c937ce0301734cac207a02b656f46) and also for [the stock status counts](https://github.com/woocommerce/woocommerce-blocks/pull/8599/commits/f947d2202dab6613a33678ae46c3351f71b3586d).

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (nefeline) [reply src/StoreApi/Utilities/ProductQueryFilters.php]: > Interesting. When I tested, if I use 2.5, I don't get 2 also.

Hm, I'm not sure if I follow: out of curiosity, how would you re-write this query with a `BETWEEN` and also circumvent the problem of having the same product counted multiple times for different intervals?

Using your query as an example on a shop with a product with average rating = 2.5 and another one with average rating = 3.5:

```sql
select distinct product_id, average_rating from wp_wc_product_meta_lookup where average_rating BETWEEN 1.5 and 2.5;
select distinct product_id, average_rating from wp_wc_product_meta_lookup where average_rating BETWEEN 2.5 and 3.5;
select distinct product_id, average_rating from wp_wc_product_meta_lookup where average_rating BETWEEN 3.5 and 4.5;
```

Results for the first query:
Screenshot 2023-04-05 at 16 24 27

Results for the second query:
Screenshot 2023-04-05 at 16 24 58

Results for the third query:
Screenshot 2023-04-05 at 16 26 12


In this scenario, product ID 15 comes up on the count results for the first and the second queries. Also, product ID 12 comes up on the count results for the second and third queries, when they shouldn't, as this way the counts are duplicated/incorrect.

Comment (roykho) [reply src/StoreApi/Utilities/ProductQueryFilters.php]: Perhaps we could `array_unique` on the result on every loop so it will not contain any duplicates. Just a thought as an option.

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (nefeline) [reply src/StoreApi/Utilities/ProductQueryFilters.php:122]: > I think it would still cause an issue because you have changed the function signature.

Good point, the second param was modified from an `$attributes` array to a `$filtered_attribute` string. How about keeping things contained and deprecating [the modified second param](https://github.com/woocommerce/woocommerce-blocks/pull/8599/commits/640f66fd4828b559afe36694d0bc092fe3250b7e)?

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (nefeline) [reply src/StoreApi/Utilities/ProductQueryFilters.php]: > Perhaps we could array_unique on the result on every loop so it will not contain any duplicates. Just a thought as an option.

Hmm, but that would not solve the problem either. How would we define the correct count for each one of the results via PHP? In our example here, how to determine if Product ID 15 belongs to the first or the second queries after they were already fetched from the DB? Also, why would we prefer to fetch incorrect results from the DB to work on their logic later via PHP (assuming there's a way to do so reliably)?

The counts are working as expected the way they are structured now: I'm glad to adopt alternative solutions if they are better in terms of efficiency/performance, but I'm not seeing the usage of `BETWEEN` as a way to achieve this yet. How about pushing these changes to a separate branch to compare both approaches?

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (roykho) [inline src/StoreApi/Utilities/ProductQueryFilters.php]: For the caching, what do you think about adding a check for `WP_DEBUG` constant? This way developers can test without having to comment out the `return` and prevent false negatives. It is common for developers to set `WP_DEBUG` to `true` in development environment so we could use this to our advantage especially when testing.

`if ( ! empty( $cached_results ) && ( defined( 'WP_DEBUG' ) && ! WP_DEBUG ) ) {`

Comment (nefeline) [reply src/StoreApi/Utilities/ProductQueryFilters.php]: Good idea! [Done](https://github.com/woocommerce/woocommerce-blocks/pull/8599/commits/497e101a5f58a7bfc82cf3415b33ab05af925a31).

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (roykho) [inline src/StoreApi/Utilities/ProductQueryFilters.php:277]: Can we format this block of query code with indentation because VSCode is showing formatting alerts.

**Review by roykho — CHANGES_REQUESTED** [review]
Left a small comment about the code formatting and the PHP Unit tests are failing on the attributes count. Besides those two, I think we're good to go.

**Comment by nefeline** [comment]
> Left a small comment about the code formatting and the PHP Unit tests are failing on the attributes count. Besides those two, I think we're good to go.

Thanks! Regarding the unit tests, those were already on my radar (mentioned in the description of this PR).

File: src/StoreApi/Utilities/ProductQueryFilters.php
Comment (nefeline) [reply src/StoreApi/Utilities/ProductQueryFilters.php:277]: Could you give an example, please? My VSCode is not showing any formatting alerts, and things are a-ok regarding coding standards (PHPCS is not throwing any errors); maybe that's coming from an extension?

Comment (roykho) [reply src/StoreApi/Utilities/ProductQueryFilters.php:277]: Yeah I think it has mixed of tabs and spaces.

Comment (nefeline) [reply src/StoreApi/Utilities/ProductQueryFilters.php:277]: Gotcha: I'm a bit in the dark here as none of my IDEs are flagging this problem, but pushed this formatting update just in case  https://github.com/woocommerce/woocommerce-blocks/pull/8599/commits/cb40f042651b931875c652dc5136c6e4797a3bd8

Comment (roykho) [reply src/StoreApi/Utilities/ProductQueryFilters.php:277]: I pushed up an update to fix that. Thanks!

**Review by roykho — APPROVED** [review]
Thanks for all the fixes. Everything tested nicely per description. Since the unit tests will be done in a separate PR, I will go ahead and approve this for now.

**Comment by nefeline** [comment]
> Thanks for all the fixes. Everything tested nicely per description. Since the unit tests will be done in a separate PR, I will go ahead and approve this for now.

Excellent! Thanks for the suggestions regarding caching and spotting the glitch when more than one product meta was selected: it's great to have this out of the way sooner than later. 

https://github.com/woocommerce/woocommerce-blocks/pull/8984 was created to address the PHP Unit tests: Both PRs will be merged once the tests are approved and I also get a second review on this one.

**Comment by opr** [comment]
Bumping to 10.1.0 milestone after p1681128984753939/1681128388.500929-slack-C02UBB1EPEF

**Review by thealexandrelara — APPROVED** [review]
Thank you for working on this, I tested with different combinations of filters and it is working as expected. Great job! 🚀

**Comment by kmanijak** [comment]
I know it's been merged, but while testing the 10.1.0 I was wondering if changing counters are expected:

> Should we update the counts when filtering attributes with the condition = any ( or )? Currently, the counts are updated to match the filtered product results. Alternatively, we could keep the counts unchanged.

In my opinion, it would be better to keep the counts unchanged because as a user I feel like I'm getting lost in what the number actually represents. Quite easily we can make the same filter show multiple numbers:
- default number,
- zero if there are no products after filtering another attribute,
- something in between when there's a combination of at least two attributes.

**Comment by roykho** [comment]
Hi @kmanijak - here is actually the more detailed public discussion on this topic https://github.com/woocommerce/woocommerce-blocks/discussions/9023

**Comment by fabianmarz** [comment]
Hi @roykho, @nefeline, I stumbled upon this PR as I'm about to upgrade WooCommerce and noticed a breaking change on my end. I'm using the [collection-data](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/StoreApi/docs/product-collection-data.md) API to get a pre-filtered result set based on the some categories. This doesn't seem to work properly anymore since this PR dropped `ProductQuery()` support.

Here's a screenshot of my `WP_REST_Request` instance but the request returns all attributes instead of the ones filtered for the given category terms
image

I noticed the Collection API docs still state support for all the parameters
> In addition to the above attributes, all product list attributes are supported. This allows you to get data for a certain subset of products. See [the products API list products section](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/StoreApi/docs/products.md#list-products) for the full list.

Any help would be appreciated here Thank you! 🙌

**Comment by fabianmarz** [comment]
I'm still struggling updating to the recent WC version since this change broke the collection-data API endpoint mentioned here https://github.com/woocommerce/woocommerce/issues/42157.

Could someone please give me some guidance how I can get a subset of products based on a category and use the filters?

Thanks!

**Comment by creative-andrew** [comment]
@nefeline  @roykho do we have any news on this? This PR introduced a breaking change that invalidates the following:

> In addition to the above attributes, all product list attributes are supported. This allows you to get data for a certain subset of products. See [the products API list products section](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/StoreApi/docs/products.md#list-products) for the full list.

**Pull request merged**
